src/gwt/panmirror/src/editor/src/behaviors/spelling/spelling-realtime.ts

```diff
@@ -209,28 +218,55 @@ function spellingDecorations(
   from = -1,
   to = -1,
 ): Decoration[] {
+
+  // a map of wordText -> [wordRange...]
+  const rangeMap = new Map<string, EditorWordRange[]>();
+
   // break words
   const words = getWords(state, from, to, spelling, excluded);
 
   // spell check and return decorations for misspellings
   const decorations: Decoration[] = [];
+
+  // words to pass to the spellchecker
+  const wordsToCheck: string[] = [];
+
   while (words.hasNext()) {
     const word = words.next()!;
     const wordText = state.doc.textBetween(word.start, word.end);
-    const wordCheck = spellcheckerWord(wordText);
-    if (!spelling.checkWord(wordCheck)) {
-      const attrs: DecorationAttrs = {};
-      const spec: { [key: string]: any } = {
-        word: wordCheck,
-      };
-      if (excludeCursor && state.selection.head > word.start && state.selection.head <= word.end) {
-        spec.cursor = true;
-      } else {
-        attrs.class = kSpellingErrorClass;
+
+    let ranges = rangeMap.get(wordText);
```

Comment:
```suggestion
    let ranges = rangeMap.get(wordText) || [];
```

IIUC this would avoid the need for the `!!ranges` check below? (Or does this pattern make the Typescript type checker unhappy?)

Reply:
(this is purely just a style comment so feel free to disregard if appropriate)

Reply:
This appears to work! I like it.